Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -342,6 +343,295 @@ lemma exists_Solution_of_Solution' : ∃ (S₁ : Solution hζ), S₁.multiplicit
+
+/-- If `p : 𝓞 K` is a prime that divides both `S.a + S.b` and `S.a + η * S.b`, then `p`
+is associated with `λ`. -/
+lemma associated_of_dvd_a_add_b_of_dvd_a_add_eta_mul_b {p : 𝓞 K} (hp : Prime p)
+    (hpab : p ∣ S.a + S.b) (hpaetab : p ∣ S.a + η * S.b) : Associated p λ := by
+  by_cases p_lam : (p ∣ λ)
+  · exact hp.associated_of_dvd hζ.zeta_sub_one_prime' p_lam
+  exfalso
+  refine hp.not_unit <| IsCoprime.isUnit_of_dvd' S.coprime ?_ ?_
+  · refine (Prime.dvd_or_dvd hp ?_).resolve_left ‹_›
+    rw [show λ * S.a = η * (S.a + S.b) - (S.a + η * S.b) by rw [coe_eta]; ring]
+    exact dvd_sub (dvd_mul_of_dvd_right hpab _) hpaetab
+  · refine (Prime.dvd_or_dvd hp ?_).resolve_left ‹_›
+    rw [show λ * S.b = (S.a + η * S.b) - (S.a + S.b) by rw [coe_eta]; ring]
+    exact dvd_sub hpaetab hpab
+
+/-- If `p : 𝓞 K` is a prime that divides both `S.a + S.b` and `S.a + η ^ 2 * S.b`, then `p`
+is associated with `λ`. -/
+lemma associated_of_dvd_a_add_b_of_dvd_a_add_eta_sq_mul_b {p : 𝓞 K} (hp : Prime p)
+    (hpab : p ∣ (S.a + S.b)) (hpaetasqb : p ∣ (S.a + η ^ 2 * S.b)) : Associated p λ := by
+  by_cases p_lam : (p ∣ λ)
+  · exact hp.associated_of_dvd hζ.zeta_sub_one_prime' p_lam
+  exfalso
+  refine hp.not_unit <| S.coprime.isUnit_of_dvd' ?_ ?_
+  · refine (hp.dvd_or_dvd ?_).resolve_left p_lam
+    rw [show λ * S.a = - (1 - η) * S.a by rw [coe_eta]; ring, ← hζ.toInteger_cube_eq_one]
+    rw [show - (hζ.toInteger ^ 3 - η) * S.a = η * ((S.a + η ^ 2 * S.b) - η ^ 2 * (S.a + S.b))
+      by rw [coe_eta]; ring, (Units.isUnit η).dvd_mul_left]
+    exact hpaetasqb.sub (dvd_mul_of_dvd_right hpab _)
+  · refine (hp.dvd_or_dvd ?_).resolve_left p_lam
+    rw [show λ * S.b = - (1 - η) * S.b by rw [coe_eta]; ring, ← hζ.toInteger_cube_eq_one]
+    rw [show - (hζ.toInteger ^ 3 - η) * S.b = η * ((S.a + S.b) - (S.a + η ^ 2 * S.b))
+      by rw [coe_eta]; ring, (Units.isUnit η).dvd_mul_left]
+    exact hpab.sub hpaetasqb
+
+/-- If `p : 𝓞 K` is a prime that divides both `S.a + η * S.b` and `S.a + η ^ 2 * S.b`, then `p`
+is associated with `λ`. -/
+lemma associated_of_dvd_a_add_eta_mul_b_of_dvd_a_add_eta_sq_mul_b {p : 𝓞 K} (hp : Prime p)
+    (hpaetab : p ∣ S.a + η * S.b) (hpaetasqb : p ∣ S.a + η ^ 2 * S.b) : Associated p λ := by
+  by_cases p_lam : (p ∣ λ)
+  · exact hp.associated_of_dvd hζ.zeta_sub_one_prime' p_lam
+  exfalso
+  refine hp.not_unit <| S.coprime.isUnit_of_dvd' ?_ ?_
+  · refine (hp.dvd_or_dvd ?_).resolve_left p_lam
+    rw [show λ * S.a = η * (S.a + η * S.b) - (S.a + η ^ 2 * S.b) by rw [coe_eta]; ring]
+    exact dvd_mul_of_dvd_right hpaetab _ |>.sub hpaetasqb
+  · refine (hp.dvd_or_dvd ?_).resolve_left p_lam
+    rw [← (Units.isUnit η).dvd_mul_left, show η * (λ * S.b) = (S.a + η ^ 2 * S.b) - (S.a + η * S.b)
+      by rw [coe_eta]; ring]
+    exact hpaetasqb.sub hpaetab
+
+/-- Given `S : Solution`, we let `S.y` be any element such that `S.a + η * S.b = λ * S.y` -/
+noncomputable def y := (lambda_dvd_a_add_eta_mul_b S).choose
```

Comment:
All the various `y`, `z`, `w` etc are mathematically arbitrary elements that satisfy the `spec` lemma. Technically, using `choose_spec` we fix one single elements, and all the proofs are wrt this fixed elements, not an arbitrary one. Usually this is a bad idea, but here I think it is OK:
* the notation is nicer (we don't have to keep `{y} (hy)` around
* All the lemmas are useful only to prove flt3, and will never be used for something else.

Reply:
Sounds like they should be private

Reply:
Ah sure. Do you know if there is way to make a whole section private?

Reply:
It [seems](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/private.20section) there is no simple way.